File: 951-flip-equivalent-binary-trees/flip-equivalent-binary-trees.c

bool flipEquiv(struct TreeNode* root1, struct TreeNode* root2) {
    // If both nodes are NULL, they are equivalent
    if (root1 == NULL && root2 == NULL) {
        return true;
    }
    
    // If one node is NULL and other isn't, they are not equivalent
    if (root1 == NULL || root2 == NULL) {
        return false;
    }
    
    // If values are different, they are not equivalent
    if (root1->val != root2->val) {
        return false;
    }
    
    // Check both possible arrangements:
    // 1. Without flip: left with left, right with right
    // 2. With flip: left with right, right with left
    return (flipEquiv(root1->left, root2->left) && flipEquiv(root1->right, root2->right)) ||
           (flipEquiv(root1->left, root2->right) && flipEquiv(root1->right, root2->left));
}

File: 988-flip-equivalent-binary-trees/flip-equivalent-binary-trees.java
class Solution {
    public boolean flipEquiv(TreeNode root1, TreeNode root2) {
        // If both nodes are null, they are equivalent
        if (root1 == null && root2 == null) {
            return true;
        }
        
        // If one node is null and other isn't, they are not equivalent
        if (root1 == null || root2 == null) {
            return false;
        }
        
        // If values are different, they are not equivalent
        if (root1.val != root2.val) {
            return false;
        }
        
        // Check both possible arrangements:
        // 1. Without flip: left with left, right with right
        // 2. With flip: left with right, right with left
        return (flipEquiv(root1.left, root2.left) && flipEquiv(root1.right, root2.right)) ||
               (flipEquiv(root1.left, root2.right) && flipEquiv(root1.right, root2.left));
    }
}